### ISSUE TITLE
Render FPS takes a nose dive to maintain 60 physics FPS at all costs. 

### ISSUE THREAD

**Issue by BlackMoons** [Regalis11/Barotrauma#12860]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?

When using a complicated ship (3600 items), docked at a station with NPC's (first station of the campaign), using a complex mod on a low end PC (3.2ghz CPU + 1060GTX) results in excessive drop in render FPS to maintain 60 physics FPS. 

Screenshots:

Neurotrauma + complicated ship, Single player campaign, first outpost (15 render FPS, 60 physics FPS, Render time 8.9mS, update avg 14ms)

https://cdn.discordapp.com/attachments/953553613194682368/1164727029242019912/image.png?ex=65444393&is=6531ce93&hm=6a82c979617a29859d793a4b20d4edf5cd914ad68dde7d61753ba78ffc8a48c5&

Same ship, but without neurotrauma (38 render FPS, 60 physics FPS, Render avg time 8.3ms, Update avg 11ms)

https://cdn.discordapp.com/attachments/953553613194682368/1164727569715843152/image.png?ex=65444414&is=6531cf14&hm=1d39ee8b70f118b2639617a86e700da464366fabf3c585b3cc78f4cbc0cd696b&

I suspect the engine is prioritizing physics FPS over render FPS. An additional 3mS to update time shouldn't cost over half the render FPS. 

Trying performance fix + neurotrauma, with entity update frequency set to 1, and 'timing accumulator max' to 50

https://cdn.discordapp.com/attachments/888972313977028638/1164981905213632593/image.png?ex=654530f2&is=6532bbf2&hm=c60baf75ff2053325dab32c14c6c16808a8ec0b76112fd38749e72c289534212&

Results in same update time (14mS), but now it has 37FPS and 50 physics FPS and is actually playable. (Although running at 80% real time speed)

(Entity update frequency 1, Timing accumulator 250, what the mod states is barotrauma's default, results in similar FPS as without performance fix) 

Similar performance loss was also observed using the ship upon reaching the 2nd outpost with other mods without neurotrauma or other LUA mods in a campaign game, where performance suddenly increased after accepting missions (Likely just a tiny change in update time caused a doubling of FPS)

Attached below is the submarine. 

[Complex Sub.zip](https://github.com/Regalis11/Barotrauma/files/13057192/Complex.Sub.zip)

Suggestion: Lower the accumulator setting, and allow the server to have a 'Kick if client slowing server below X% realtime' setting. (People with friends likely want this set really low or turned off, while people running a public server might want it set 80%+ unless they are running a really complex ship with a ton of mods)

Or at least, lower the setting if FPS drop below 30 or something like that.

Any suggestions on optimizing my ship (without removing massive amounts of functionality) would also be appreciated. Or at least point out what functionality is eating the most CPU, since even deleting a lot of logic seems to have very little impact on its update time. (other then deleting every relay on the ship, that had an impact, and disabled every bit of logic on the ship nearly, No other component had nearly as much impact. Motion sensors and delay gates hardly added anything) 

### Reproduction steps

Overload barotrauma with entity/item updates until render FPS drop below 60FPS. and/or play it on a low end PC.

Then overload it a little more and watch render FPS dive incredibly fast while it tries to maintain 60 physics FPS. 

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Happens in both single player and multiplayer

### Version

v1.1.18.1 (Treacherous Tides)

### Which operating system did you encounter this bug on?

Windows

**Comment by Regalis11** [comment]
Thank you for the report! I think this somewhat falls under the [General performance problems](https://github.com/Regalis11/Barotrauma/issues/9396) ticket, but there's some additional info here so I think we can keep it open for now.

> Suggestion: Lower the accumulator setting, and allow the server to have a 'Kick if client slowing server below X% realtime' setting. (People with friends likely want this set really low or turned off, while people running a public server might want it set 80%+ unless they are running a really complex ship with a ton of mods)

I'm not entirely sure what you mean by lowering the "accumulator setting". Do you mean allowing the game to run at a physics update rate of less than 60 updates per second? I'm also not sure what you mean by "slowing server below X% realtime" - do you mean the server would try to match the update rate of the client whose game is performing the worst?

**Comment by BlackMoons** [comment]
Yes, allowing the game to run at less then 60 updates per second. It slows the game down (slow motion) since it seems to be a fixed timestep. (I understand variable timestep tends to ruin any determinism) 

The server should be already matching the 'slowest' person on the server, but some people might not want that (on public servers) and would rather they be kicked if they make the game too slow. (Presumably, massive rubber banding or something else bad would happen if some client or the server was not updating physics at the same speed as everyone else)

**Comment by BlackMoons** [comment]
PS: the accumulator is what performance fix calls it. Maybe a quick check on the mod to see what its actually altering? It seems to be a setting related to how much time it will spend on physics vs rendering when it can't maintain 60physics FPS.

**Comment by BlackMoons** [comment]
Also, looking at the screenshots in https://github.com/Regalis11/Barotrauma/issues/9396

The first shows 62 physics FPS at an average update time of 14.69ms

Meaning 0.91 of every second (62 * 14.69) was spent on physics. (91%) 

Leaving 0.09 seconds every second for rendering. at 10mS avg render time, that results in just 9FPS (11FPS shown in the screenshot, so very close to what my math says)

If you allowed even a 20% slowdown (50physics FPS), that would result in 0.734 cpu time spent on physics, resulting in 26~FPS.

And if you switch the entire game to 30FPS physics/electrical/etc update like many games use (interpolating the 30fps physics results to achieve 60+FPS motion is common and low CPU usage, and even reduces temporal aliasing since you can now run the display itself at say, 40 or 50fps and interpolate for the exact time between frames), it would be 0.44s spent on physics, with a resulting 66FPS~ on the exact same system that was getting 11FPS with 60FPS physics. (Maybe 50~60FPS after overhead of interpolation)

**Comment by Regalis11** [comment]
> The first shows 62 physics FPS at an average update time of 14.69ms
> Meaning 0.91 of every second (62 * 14.69) was spent on physics. (91%)
> Leaving 0.09 seconds every second for rendering. at 10mS avg render time, that results in just 9FPS (11FPS shown in the screenshot, so very close to what my math says)

I think you're somewhat misunderstanding how the physics update rate works. The physics FPS being 62 in that pic doesn't tell you anything about how long updating the physics took - the game just aims to run the physics update 60 times per second, no matter if one of those updates takes 0.1 ms or 10 ms. And actually, if you look at the actual physics update time in that pic (3.17 ms), and assume the previous frame took about the average amount of time, that would mean 3.17 * 60 = 190.2 ms (19 %) was spent on physics during the previous second.

The situation seems to be similar in the screenshots you posted too: in the initial screenshot for example, the previous physics update took 0.98 ms, which means only about 5.8% of the previous second was spent on calculating physics.

> The server should be already matching the 'slowest' person on the server, but some people might not want that (on public servers) and would rather they be kicked if they make the game too slow.

In my opinion that's absolutely not how it should work. People who can run the game well shouldn't be restricted by one client who can't.

> PS: the accumulator is what performance fix calls it. Maybe a quick check on the mod to see what its actually altering? It seems to be a setting related to how much time it will spend on physics vs rendering when it can't maintain 60 render FPS.

I'm pretty familiar with how Performance Fix works (we've even taken some inspiration for vanilla optimizations from it), and to my knowledge, it doesn't touch the physics update rate at all. It only makes certain things, like items, hulls and gaps, be updated less frequently. That approach unfortunately causes quite a lot of issues, because a lot of the logic in items for example has been written under the assumption that it runs every frame, and as a consequence Performance Fix tends to cause various bugs.

**Comment by Regalis11** [comment]
But regardless, we could consider dropping updates more aggressively when the game is struggling to maintain the fixed 60 updates per second. Currently it only starts slowing the updates down if the performance gets extremely bad (more specifically, if it's 25 milliseconds "behind" in how much time it needs to simulate).

I think this ticket isn't quite a bug report though, more of a suggestion on how bad performance could potentially be addressed, so I think it should be moved to discussions. Thank you for the ticket regardless, this is definitely be something worth investigating, and I think I'll also give the custom sub a test in a profiler to see what's eating up performance there.
